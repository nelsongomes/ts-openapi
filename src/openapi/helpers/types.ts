import Joi from "joi";

type Common = {
  description?: string;
  required?: boolean;
  nullable?: boolean;
};

type ParameterCommon = { isParameter?: boolean };
type ObjectCommon = { modelName?: string };

type StringParameters = {
  default?: string;
  example?: string;
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;

type StringEnumParameters = {
  values: string[];
  default?: string;
  example?: string;
} & Common &
  ParameterCommon;

type EmailParameters = StringParameters;
type UriParameters = StringParameters;
type HostnameParameters = StringParameters;
type PasswordParameters = {
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;
type UuidParameters = {
  example?: string;
} & Common &
  ParameterCommon;
type IpParameters = {
  example?: string;
} & Common &
  ParameterCommon;
type BinaryParameters = StringParameters;
type ByteParameters = StringParameters;

type DateParameters = {
  default?: string;
  example?: string;
} & Common &
  ParameterCommon;

type IntegerParameters = {
  default?: number;
  example?: number;
  minValue?: number;
  maxValue?: number;
} & Common &
  ParameterCommon;

type IntegerEnumParameters = {
  values: any[];
  default?: number;
  example?: number;
} & Common &
  ParameterCommon;

type NumberEnumParameters = IntegerEnumParameters & ParameterCommon;

type BooleanParameters = {
  default?: boolean;
  example?: boolean;
} & Common &
  ParameterCommon;

type ObjectParameters = {
  properties: Joi.SchemaMap<any>;
  default?: object;
  example?: object;
} & Common &
  ObjectCommon;

type ArrayParameters = {
  arrayType: Joi.SchemaLikeWithoutArray;
  default?: object;
  example?: object;
  minLength?: number;
  maxLength?: number;
} & Common &
  ParameterCommon;

export const Types = {
  String: (parameters?: StringParameters) => {
    let joiType = Joi.string();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        maxLength,
        minLength,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (minLength) {
        joiType = joiType.min(minLength);
      }

      if (maxLength) {
        joiType = joiType.max(maxLength);
      }

      if (parameters.default) {
        joiType = joiType.default(parameters.default);
      }

      if (parameters.example) {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  StringEnum: (parameters: StringEnumParameters) => {
    return Types.String(parameters).valid(...parameters.values);
  },

  Email: (parameters?: EmailParameters) => {
    return Types.String(parameters).email();
  },

  Password: (parameters?: PasswordParameters) => {
    return Types.String(parameters).meta({ format: "password" });
  },

  Uuid: (parameters?: UuidParameters) => {
    return Types.String(parameters)
      .uuid({ version: "uuidv4" })
      .meta({ format: "uuid" })
      .min(36)
      .max(36);
  },

  Uri: (parameters?: UriParameters) => {
    return Types.String(parameters)
      .uri()
      .meta({ format: "uri" });
  },

  Hostname: (parameters?: HostnameParameters) => {
    return Types.String(parameters)
      .hostname()
      .meta({ format: "hostname" });
  },

  Ipv4: (parameters?: IpParameters) => {
    return Types.String(parameters)
      .ip({ version: ["ipv4"] })
      .meta({ format: "ipv4" })
      .min(7)
      .max(15);
  },

  Ipv6: (parameters?: IpParameters) => {
    return Types.String(parameters)
      .ip({ version: ["ipv6"] })
      .meta({ format: "ipv6" })
      .min(3)
      .max(45);
  },

  Binary: (parameters?: BinaryParameters) => {
    let joiType = Joi.binary();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        maxLength,
        minLength,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (minLength) {
        joiType = joiType.min(minLength);
      }

      if (maxLength) {
        joiType = joiType.max(maxLength);
      }

      if (parameters.default) {
        joiType = joiType.default(parameters.default);
      }

      if (parameters.example) {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Byte: (parameters?: ByteParameters) => {
    return Types.Binary(parameters).encoding("base64");
  },

  DateTime: (parameters?: DateParameters) => {
    let joiType = Joi.string()
      .isoDate()
      .max(24);

    if (parameters) {
      const { description, required, nullable } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Date: (parameters?: DateParameters) => {
    return Types.DateTime(parameters)
      .meta({ format: "date" })
      .min(10)
      .max(10);
  },

  Number: (parameters?: IntegerParameters) => {
    let joiType = Joi.number();

    if (parameters) {
      const {
        description,
        required,
        nullable,
        minValue,
        maxValue,
        example
      } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean" && required) {
        joiType = joiType.required();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (typeof minValue === "number") {
        joiType = joiType.min(minValue);
      }

      if (typeof maxValue === "number") {
        joiType = joiType.max(maxValue);
      }

      if (typeof parameters.default === "number") {
        joiType = joiType.default(parameters.default);
      }

      if (typeof parameters.example === "number") {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  NumberEnum: (parameters: NumberEnumParameters) => {
    return Types.Number(parameters).valid(...parameters.values);
  },

  Integer: (parameters?: IntegerParameters) => {
    return Types.Number(parameters).integer();
  },

  IntegerEnum: (parameters: IntegerEnumParameters) => {
    return Types.Integer(parameters).valid(...parameters.values);
  },

  Boolean: (parameters?: BooleanParameters) => {
    let joiType = Joi.boolean().strict();

    if (parameters) {
      const { description, required, nullable, example } = parameters;

      if (description) {
        joiType = joiType.description(description);
      }

      if (typeof required === "boolean") {
        joiType = required ? joiType.required() : joiType.optional();
      }

      if (typeof nullable === "boolean" && nullable) {
        joiType = joiType.allow(null);
      }

      if (typeof parameters.default === "boolean") {
        joiType = joiType.default(parameters.default);
      }

      if (typeof parameters.example === "boolean") {
        joiType = joiType.example(example);
      }

      if (parameters.isParameter) {
        joiType = joiType.meta({ parameter: true });
      }
    }

    return joiType;
  },

  Object: (parameters: ObjectParameters) => {
    const { properties, description, required, nullable, example } = parameters;
    let joiType = Joi.object().keys(properties);

    if (description) {
      joiType = joiType.description(description);
    }

    if (typeof required === "boolean") {
      joiType = required ? joiType.required() : joiType.optional();
    }

    if (nullable) {
      joiType = joiType.allow(null);
    }

    if (parameters.default) {
      joiType = joiType.default(parameters.default);
    }

    if (parameters.example) {
      joiType = joiType.example(example);
    }

    if (parameters.modelName) {
      joiType = joiType.meta({ modelName: parameters.modelName });
    }

    return joiType;
  },

  Array: (parameters: ArrayParameters) => {
    const {
      arrayType,
      description,
      required,
      nullable,
      example,
      minLength,
      maxLength
    } = parameters;
    let joiType = Joi.array().items(arrayType);

    if (description) {
      joiType = joiType.description(description);
    }

    if (typeof required === "boolean") {
      joiType = required ? joiType.required() : joiType.optional();
    }

    if (nullable) {
      joiType = joiType.allow(null);
    }

    if (parameters.default) {
      joiType = joiType.default(parameters.default);
    }

    if (parameters.example) {
      joiType = joiType.example(example);
    }

    if (minLength) {
      joiType = joiType.min(minLength);
    }

    if (maxLength) {
      joiType = joiType.max(maxLength);
    }

    if (parameters.isParameter) {
      joiType = joiType.meta({ parameter: true });
    }

    return joiType;
  }
};
