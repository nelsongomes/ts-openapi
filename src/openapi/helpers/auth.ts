import {
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
  SecurityOauth2Scheme,
  SecurityOauth2Scopes,
} from "../openapi.types";

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}

export function oauth2ImplicitAuth(
  description: string,
  authorizationUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      implicit: {
        authorizationUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2AuthorizationCodeAuth(
  description: string,
  authorizationUrl: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      authorizationCode: {
        authorizationUrl,
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2PasswordAuth(
  description: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      password: {
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}

export function oauth2ClientCredentialsAuth(
  description: string,
  tokenUrl: string,
  scopes: SecurityOauth2Scopes,
  refreshUrl?: string
): SecurityOauth2Scheme {
  return {
    type: "oauth2",
    description,
    flows: {
      clientCredentials: {
        tokenUrl,
        scopes,
        refreshUrl,
      },
    },
  };
}
