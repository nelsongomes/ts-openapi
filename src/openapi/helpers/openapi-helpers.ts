import {
  Body,
  SecurityApiKeyScheme,
  SecurityApiKeySchemeMethod,
  SecurityBasicScheme,
  SecurityBearerScheme,
} from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}

export function basicAuth(): SecurityBasicScheme {
  return {
    type: "http",
    scheme: "basic",
  };
}

export function apiKeyAuth(
  name: string,
  method: SecurityApiKeySchemeMethod
): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: method,
    name,
  };
}

export function cookieAuth(cookieName: string): SecurityApiKeyScheme {
  return {
    type: "apiKey",
    in: "cookie",
    name: cookieName,
  };
}

export function bearerAuth(
  bearerFormat?: "JWT" | string
): SecurityBearerScheme {
  return {
    type: "http",
    scheme: "bearer",
    ...(bearerFormat && { bearerFormat }),
  };
}
