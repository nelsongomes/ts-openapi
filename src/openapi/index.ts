import Joi, { Schema } from "joi";

export { Joi, Schema };

export { OpenApi } from "./openapi";
export { WebRequestSchema, ParameterIn, Parameters } from "./openapi.types";

export { textPlain } from "./openapi-helpers";
