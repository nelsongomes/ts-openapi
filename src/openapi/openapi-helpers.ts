export function textPlain(description: string) {
  return {
    content: {
      "text/plain": {
        schema: {
          type: "string",
        },
      },
    },
    description,
  };
}
