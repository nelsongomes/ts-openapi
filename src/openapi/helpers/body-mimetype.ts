import { Body } from "../openapi.types";

export function textPlain(description: string): Body {
  return {
    description,
    content: { "text-plain": {} },
  };
}
