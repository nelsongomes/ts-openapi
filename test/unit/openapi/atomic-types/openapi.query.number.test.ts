import Joi from "joi";
import { OpenApi } from "../../../../src/openapi/openapi";
import { textPlain } from "../../../../src/openapi/helpers/body-mimetype";
import { Types } from "../../../../src/openapi/helpers/types";

describe("src/openapi/openapi", () => {
  let openApi: OpenApi;

  describe("Number", () => {
    beforeEach(() => {
      openApi = new OpenApi(
        "1.0.0",
        "Server API",
        "Some test api",
        "nelson.ricardo.gomes@gmail.com"
      );

      openApi.setServers([{ url: "https://server.com" }]);
    });

    test("number simple", async () => {
      const query = Joi.object().keys({
        user_name: Types.Number()
      });

      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("number simple all options", async () => {
      const query = Joi.object({
        sensor: Types.Number({
          description: "Sensor value.",
          required: true,
          minValue: 0.5,
          maxValue: 100.66,
          default: 1,
          example: 33.333,
          nullable: true
        })
      }).description("ignore this");

      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("number enum", async () => {
      enum EnumValues {
        AAA = 4.1,
        BBB = 10,
        CCC = 20
      }

      const query = Joi.object({
        option: Types.NumberEnum({ values: Object.values(EnumValues) })
      });

      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("number enum min and max should be skipped (the values are already limited)", async () => {
      enum EnumValues {
        AAA = 3,
        BBB = 6,
        CCC = 9
      }

      const query = Joi.object({
        option: Types.NumberEnum({
          values: Object.values(EnumValues),
          description: "Integer options from enum"
        })
      });

      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("number enum default is part of enum", async () => {
      enum EnumValues {
        AAA = 10,
        BBB = 20,
        CCC = 30
      }

      const query = Joi.object().keys({
        option: Types.NumberEnum({
          values: Object.values(EnumValues),
          description: "Integer options from enum",
          default: 20
        })
      });

      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("number enum default is undefined because is NOT in enum", async () => {
      enum EnumValues {
        AAA = 10,
        BBB = 20,
        CCC = 30
      }

      const query = Types.Object({
        properties: {
          option: Types.NumberEnum({
            values: Object.values(EnumValues),
            description: "Integer options from enum",
            default: 12
          })
        }
      });

      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });

    test("number enum all options", async () => {
      enum EnumValues {
        AAA = 10.1,
        BBB = 20.2,
        CCC = 30.3
      }

      const query = Types.Object({
        properties: {
          sensor: Types.NumberEnum({
            values: Object.values(EnumValues),
            description: "Sensor value.",
            required: true,
            default: 10.1,
            example: 30.3,
            nullable: true
          })
        },
        description: "ignore it!"
      });
      openApi.addPath(
        "/test",
        {
          get: {
            description: "Test endpoint",
            operationId: "id",
            validationSchema: { query },
            responses: {
              200: textPlain("Successful operation.")
            },
            summary: "Server Test",
            tags: ["Internals"]
          }
        },
        true
      );
      expect(openApi.generateJson()).toMatchSnapshot();
    });
  });
});
