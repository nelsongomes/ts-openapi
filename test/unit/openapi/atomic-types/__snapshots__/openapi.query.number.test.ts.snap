// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/openapi/openapi Number number enum 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "option",
            "required": false,
            "schema": Object {
              "enum": Array [
                4.1,
                10,
                20,
              ],
              "format": "double",
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Number number enum all options 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Sensor value.",
            "example": 30.3,
            "in": "query",
            "name": "sensor",
            "required": true,
            "schema": Object {
              "default": 10.1,
              "enum": Array [
                10.1,
                20.2,
                30.3,
              ],
              "format": "double",
              "nullable": true,
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Number number enum default is part of enum 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Integer options from enum",
            "in": "query",
            "name": "option",
            "required": false,
            "schema": Object {
              "default": 20,
              "enum": Array [
                10,
                20,
                30,
              ],
              "format": "double",
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Number number enum default is undefined because is NOT in enum 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Integer options from enum",
            "in": "query",
            "name": "option",
            "required": false,
            "schema": Object {
              "default": 12,
              "enum": Array [
                10,
                20,
                30,
              ],
              "format": "double",
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Number number enum min and max should be skipped (the values are already limited) 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Integer options from enum (min:1, max:3)",
            "in": "query",
            "name": "option",
            "required": false,
            "schema": Object {
              "enum": Array [
                3,
                6,
                9,
              ],
              "format": "double",
              "maximum": 3,
              "minimum": 1,
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Number number simple 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Parameter without description.",
            "in": "query",
            "name": "user_name",
            "required": false,
            "schema": Object {
              "format": "double",
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;

exports[`src/openapi/openapi Number number simple all options 1`] = `
Object {
  "info": Object {
    "contact": Object {
      "email": "nelson.ricardo.gomes@gmail.com",
    },
    "description": "Some test api",
    "license": Object {
      "name": "Apache 2.0",
      "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "termsOfService": "http://swagger.io/terms/",
    "title": "Server API",
    "version": "1.0.0",
  },
  "openapi": "3.0.1",
  "paths": Object {
    "/health": Object {
      "get": Object {
        "description": "Service healthcheck endpoint",
        "operationId": "repeated",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Healthcheck",
        "tags": Array [
          "Internals",
        ],
      },
    },
    "/test": Object {
      "get": Object {
        "description": "Test endpoint",
        "operationId": "id",
        "parameters": Array [
          Object {
            "description": "Sensor value. (min:0.5, max:100.66)",
            "example": 33.333,
            "in": "query",
            "name": "sensor",
            "required": true,
            "schema": Object {
              "default": 1,
              "format": "double",
              "maximum": 100.66,
              "minimum": 0.5,
              "nullable": true,
              "type": "number",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "text/plain": Object {
                "schema": Object {
                  "type": "string",
                },
              },
            },
            "description": "Successful operation.",
          },
        },
        "summary": "Server Test",
        "tags": Array [
          "Internals",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "https://server.com",
    },
  ],
}
`;
